src/System.IO.Compression/src/System/IO/Compression/Crc32Helper.cs

```diff
@@ -449,11 +453,40 @@ internal static class Crc32Helper
             0x6EAB0882u, 0xA201081Cu, 0xA8C40105u, 0x646E019Bu, 0xEAE10678u,
             0x264B06E6u
         };
-
+        private static WorkerType GetCrcType()
+        {
+            try
+            {
+                // Make any P/Invoke into zlib to ensure we're able to find and use it.
+                // If we are, then use zlib.
+                Interop.zlib.zlibCompileFlags();
+                return WorkerType.ZLib;
+            }
+            catch
+            {
+                // Otherwise, fallback to managed implementation if zlib isn't available
+                Debug.Fail("zlib unavailable");
+                return WorkerType.Managed;
+            }
+        }
         // Calculate CRC based on the old CRC and the new bytes 
         // See RFC1952 for details.
         static public uint UpdateCrc32(uint crc32, byte[] buffer, int offset, int length)
         {
+            WorkerType CrcType = s_CrcType;
+            Crc32Helper _crcHelper = new Crc32Helper();
+
+            if (CrcType == WorkerType.ZLib)
+            {
+                return _crcHelper.crc32_zlib(crc32, buffer, offset, length);
+            }
+            else
+            {
+                return _crcHelper.crc32_managed(crc32, buffer, offset, length);
+            }
+        }
+        public uint crc32_managed(uint crc32, byte[] buffer, int offset, int length)
```

Comment:
Can't these be static rather than instance methods?  They should also be private, with the only exposed entry point being the existing UpdateCrc32.